flatsurf/geometry/gl2r_orbit_closure.py

```diff
@@ -1094,34 +1096,24 @@ def update_tangent_space_from_flow_decomposition(self, decomposition):
             self.update_tangent_space_from_vector(v)
             if self._U_rank == self._U.nrows(): return
 
-    def _rank(self, v):
+    def _rank(self):
         r"""
-        Return a lower bound for the rank of the matrix U.
+        Return a lower bound for the rank of the matrix _U.
         """
-        self._U[self._U_rank] = v
-
-        K = self._U.base_ring()
-        if K is QQ:
-            return self._U.rank()
-
-        k = K.residue_field(self._good_prime)
-        try:
-            self._Ubar[self._U_rank] = v.change_ring(k)
-        except ValueError:
-            # Some denominator in v is not a unit mod q.
-            while True:
-                from sage.all import next_prime
-                self._good_prime = next_prime(self._good_prime.residue_field().characteristic())
-                self._good_prime = K.ideal(self._good_prime).factor()[0][0]
-                k = K.residue_field(self._good_prime)
+        while True:
+            if self._p is not None:
                 try:
-                    self._Ubar = self._U.change_ring(k)
+                    self._Ubar = self._U.apply_map(phi=self._p.reduce, R=self._p.residue_field())
+                    break
                 except ValueError:
```

Comment:
Why does this catch exactly? I think it ought to be mentioned as a comment.

Reply:
```suggestion
                except ValueError:
                    # The matrix _U cannot be reduced mod p because some entries have negative valuation.
```